src/backend/cod4.c

```diff
@@ -3479,27 +3482,40 @@ code *cdbtst(elem *e, regm_t *pretregs)
 
     if ((retregs = (*pretregs & (ALLREGS | mBP))) != 0) // if return result in register
     {
-        code *nop = CNIL;
-        regm_t save = regcon.immed.mval;
-        code *cg = allocreg(&retregs,&reg,TYint);
-        regcon.immed.mval = save;
-        if ((*pretregs & mPSW) == 0)
+        if (tysize(e->Ety) == 1)
         {
-            cg = cat(cg,getregs(retregs));
-            cg = genregs(cg,0x19,reg,reg);              // SBB reg,reg
-            cg = gen2(cg,0xF7,modregrmx(3,3,reg));      // NEG reg
+            assert(I64 || retregs & BYTEREGS);
+            code *cg = allocreg(&retregs,&reg,TYint);
+            cg = gen2(cg,0x0F92,modregrmx(3,0,reg));        // SETC reg
+            if (I64 && reg >= 4)
+                code_orrex(cg, REX);
+            *pretregs = retregs;
+            c2 = cat(c2,cg);
         }
         else
         {
-            cg = movregconst(cg,reg,1,8);               // MOV reg,1
-            nop = gennop(nop);
-            cg = genjmp(cg,JC,FLcode,(block *) nop);    // Jtrue nop
-                                                        // MOV reg,0
-            movregconst(cg,reg,0,8);
-            regcon.immed.mval &= ~mask[reg];
+            code *nop = CNIL;
+            regm_t save = regcon.immed.mval;
+            code *cg = allocreg(&retregs,&reg,TYint);
+            regcon.immed.mval = save;
+            if ((*pretregs & mPSW) == 0)
+            {
+                cg = cat(cg,getregs(retregs));
+                cg = genregs(cg,0x19,reg,reg);              // SBB reg,reg
+                cg = gen2(cg,0xF7,modregrmx(3,3,reg));      // NEG reg
+            }
```

Comment:
The latency of SBB+NEG is 3 vs. 2 for SETcc+MOVZX, so you might want to use SETcc here as well though it requires 4 bytes for the code.